dj/api/engines.py

```diff
@@ -0,0 +1,86 @@
+"""
+Engine related APIs.
+"""
+
+from http import HTTPStatus
+from typing import List
+
+from fastapi import APIRouter, Depends, HTTPException
+from sqlalchemy.exc import NoResultFound
+from sqlmodel import Session, SQLModel, select
+
+from dj.models.engine import Engine
+from dj.utils import get_session
+
+router = APIRouter()
+
+
+class EngineInfo(SQLModel):
+    """
+    Class for engine creation
+    """
+
+    name: str
+    version: str
+
+
+def get_engine(session: Session, name: str, version: str) -> Engine:
+    """
+    Return an Engine instance given an engine name and version
+    """
+    statement = (
+        select(Engine).where(Engine.name == name).where(Engine.version == version)
+    )
+    try:
+        engine = session.exec(statement).one()
+    except NoResultFound as exc:
+        raise HTTPException(
+            status_code=HTTPStatus.NOT_FOUND,
+            detail=f"Engine not found: `{name}` version `{version}`",
+        ) from exc
+    return engine
+
+
+@router.get("/engines/", response_model=List[EngineInfo])
+def list_engines(*, session: Session = Depends(get_session)) -> List[EngineInfo]:
+    """
+    List all available engines
+    """
+    return list(session.exec(select(Engine)))
+
+
+@router.get("/engines/{name}/{version}/", response_model=EngineInfo)
+def list_engine(
+    name: str, version: str, *, session: Session = Depends(get_session)
+) -> EngineInfo:
+    """
+    Return an engine by name and version
+    """
+    return get_engine(session, name, version)
+
+
+@router.post("/engines/", response_model=EngineInfo)
+def add_engine(
```

Comment:
Hmm, do we want to enable people to add arbitrary engines as a part of the API? For example, someone could come in and add 'Spark' as an engine to a Druid catalog, but that wouldn't make much sense. I think these settings actually make more sense just as config. Alternatively, if we could lock these specific parts of the API down, that works as well.

Reply:
Good point, I like the latter option of thinking through how to lock down the API to just the server administrators. It's a pain having to restart the server whenever you want to add/remove/edit catalog related configurations and a lot of query engines like Spark and Trino are moving away from that towards a `CREATE|DROP CATALOG` command (which requires a lot of painful refactoring for them).

Reply:
+1 on convenience. Yeah, and along with server administration, we'll also have to think about how to deal with user management in general, since there isn't currently the concept of users.

Reply:
For user management, we can discuss the following two BDP approaches: authorized_managers VS read/write/admin. Added to the agenda.